sklearn/ensemble/_hist_gradient_boosting/binning.py

```diff
@@ -132,17 +163,25 @@ def fit(self, X, y=None):
     def transform(self, X):
         """Bin data X.
 
+        Missing values will be mapped to the first bin, but only if missing
+        values were encountered at fit time. Else, due to side effect of
+        comparing with NaNs (always results in False), missing values are
+        mapped to the last bin.
+        For this reason, `X` should be the fitting data, though we do not
+        enforce this. Note that the GBDT code only ever uses
+        mapper.fit_transform(), so this assumption is OK.
```

Comment:
Well, not anymore, since #13933.

That's a problem: if for a given feature, there are missing values in the validation data but not in the training data, the first bin will *not* be reserved for missing data, and the nans in the validation data will be incorrectly mapped.

I'm not sure how we should handle this. Here are a few suggestions:

1. raise an error?
2. do not bin validation data. We only bin it because it makes scoring faster, but strictly speaking, it's not needed. This is going to make the code a bit more convoluted, especially `raw_predict()`. 
3. determine **before** the train / val split whether there are nans in each feature. If there are nans in a given feature, then the first bin is allocated for missing values whether they exist in the training data. One downside is that if the training data has no missing value, then the first bin is allocated for nothing. (note for self: the code will need to be changed so that in this case the scan from right to left isn't performed, since it's useless.)

I have a preference for 3.

Feedback welcome!

Reply:
I implemented 3 in deda348